portalnet/src/events.rs

```diff
@@ -2,47 +2,114 @@ use std::str::FromStr;
+/// Messages that can be dispatched to an overlay network.
+pub enum OverlayRequest {
+    /// A TALK-REQ.
+    Talk(TalkRequest),
+    /// A forwarded event from another overlay.
+    Event(EventEnvelope),
+}
+impl From<EventEnvelope> for OverlayRequest {
+    fn from(event: EventEnvelope) -> Self {
+        OverlayRequest::Event(event)
+    }
+}
+impl From<TalkRequest> for OverlayRequest {
+    fn from(talk_req: TalkRequest) -> Self {
+        OverlayRequest::Talk(talk_req)
+    }
+}
+
 /// Main handler for portal network events
 pub struct PortalnetEvents {
     /// Receive Discv5 talk requests.
     pub talk_req_receiver: mpsc::Receiver<TalkRequest>,
-    /// Send overlay `TalkReq` to history network
-    pub history_overlay_sender: Option<mpsc::UnboundedSender<TalkRequest>>,
-    /// Send overlay `TalkReq` to state network
-    pub state_overlay_sender: Option<mpsc::UnboundedSender<TalkRequest>>,
-    /// Send overlay `TalkReq` to beacon network
-    pub beacon_overlay_sender: Option<mpsc::UnboundedSender<TalkRequest>>,
+    /// History network send & receive handles.
+    pub history_handle: OverlayHandle,
+    /// State network send & receive handles.
+    pub state_handle: OverlayHandle,
+    /// Beacon network send & receive handles.
+    pub beacon_handle: OverlayHandle,
     /// Send TalkReq events with "utp" protocol id to `UtpListener`
     pub utp_talk_reqs: mpsc::UnboundedSender<TalkRequest>,
 }
 
 impl PortalnetEvents {
     pub async fn new(
         talk_req_receiver: mpsc::Receiver<TalkRequest>,
-        history_overlay_sender: Option<mpsc::UnboundedSender<TalkRequest>>,
-        state_overlay_sender: Option<mpsc::UnboundedSender<TalkRequest>>,
-        beacon_overlay_sender: Option<mpsc::UnboundedSender<TalkRequest>>,
+        history_channels: OverlayChannels,
+        state_channels: OverlayChannels,
+        beacon_channels: OverlayChannels,
         utp_talk_reqs: mpsc::UnboundedSender<TalkRequest>,
     ) -> Self {
         Self {
             talk_req_receiver,
-            history_overlay_sender,
-            state_overlay_sender,
-            beacon_overlay_sender,
+            history_handle: history_channels.into(),
+            state_handle: state_channels.into(),
+            beacon_handle: beacon_channels.into(),
             utp_talk_reqs,
         }
     }
 
     /// Main loop to dispatch `Discv5` and uTP events
     pub async fn start(mut self) {
-        while let Some(talk_req) = self.talk_req_receiver.recv().await {
-            self.dispatch_discv5_talk_req(talk_req);
+        let mut receivers = vec![];
+        if let Some(rx) = self.history_handle.rx.take() {
+            receivers.push(rx);
+        }
+        if let Some(rx) = self.state_handle.rx.take() {
+            receivers.push(rx);
+        }
+        if let Some(rx) = self.beacon_handle.rx.take() {
+            receivers.push(rx);
+        }
+
+        if receivers.is_empty() {
+            warn!("Fused stream has zero receivers");
```

Comment:
I'm unsure. Prior to this, there was no guard against running the trin binary without specifying any network in the config. It might be worth it to add that now?